### PR TITLE
fix(compiler): add equiv & disp attributes to Xliff2 ICU placeholders

### PR DESCRIPTION
**Pull request by ocombe**
## PR Type
What kind of change does this PR introduce?
```
[x] Bugfix
```

## What is the current behavior?
We didn't add the equiv & disp parameters to xliff 2 ICU placeholders. This generated an error when we were loading xliff 2 translations (by the way I'm not sure why we trigger an error in this case, since ["equiv" is optional](http://docs.oasis-open.org/xliff/xliff-core/v2.0/os/xliff-core-v2.0-os.html#equiv)).

Issue Number: #17344

## What is the new behavior?
Xliff 2 ICU placeholders now have equiv & disp parameters.
For disp I chose to represent ICU expressions as `{count, plural, =0 {...} =other {...}}`. I didn't want to write the complete ICU expression since it can be nested, what do you think @vicb? I can change it for something else if you prefer.
I also added tests for ICU placeholders for xmb/xliff/xliff2 because I noticed that we didn't have any.

## Does this PR introduce a breaking change?
```
[x] No
```

### CODE REVIEW

File: packages/compiler/src/i18n/serializers/xliff2.ts [hunk 1]
```diff
@@ -173,7 +173,12 @@ class _WriteVisitor implements i18n.Visitor {
   }
 
   visitIcuPlaceholder(ph: i18n.IcuPlaceholder, context?: any): xml.Node[] {
-    return [new xml.Tag(_PLACEHOLDER_TAG, {id: (this._nextPlaceholderId++).toString()})];
+    return [new xml.Tag(_PLACEHOLDER_TAG, {
+      id: (this._nextPlaceholderId++).toString(),
```
Comment (vicb) [inline packages/compiler/src/i18n/serializers/xliff2.ts]: move `this._nextPlaceholderId++` before the return

File: packages/compiler/src/i18n/serializers/xliff2.ts [hunk 2]
```diff
@@ -173,7 +173,12 @@ class _WriteVisitor implements i18n.Visitor {
...
+      equiv: ph.name,
+      disp:
+          `{${ph.value.expression}, ${ph.value.type}, ${Object.keys(ph.value.cases).map((value: string) => value + ' {...}').join(' ')}}`
```
Comment (vicb) [inline packages/compiler/src/i18n/serializers/xliff2.ts]: Not a big fan of long expressions in interpolation (`${Object.keys(ph.value.cases).map((value: string) => value + ' {...}').join(' ')}`) consider computing the str before

File: packages/compiler/src/i18n/serializers/xliff2.ts [hunk 3]
```diff
@@ -173,7 +173,12 @@ class _WriteVisitor implements i18n.Visitor {
   }
 
   visitIcuPlaceholder(ph: i18n.IcuPlaceholder, context?: any): xml.Node[] {
-    return [new xml.Tag(_PLACEHOLDER_TAG, {id: (this._nextPlaceholderId++).toString()})];
+    return [new xml.Tag(_PLACEHOLDER_TAG, {
+      id: (this._nextPlaceholderId++).toString(),
+      equiv: ph.name,
```
Comment (vicb) [inline packages/compiler/src/i18n/serializers/xliff2.ts]: `equiv` is optional and adds no info here. Remove.

Comment (ocombe) [inline packages/compiler/src/i18n/serializers/xliff2.ts]: Well that's the point of the PR, if you read the issue, without "equiv" here, the compiler throws an error because it's missing "equiv".
Also we have "equiv" on all other placeholders for xliff2, we should have it here as well for consistency

File: packages/compiler/src/i18n/serializers/xliff2.ts (hunk 1 above)
Comment (ocombe) [inline packages/compiler/src/i18n/serializers/xliff2.ts]: it fails if I move it before the return (ids start at 2 instead of 1)

File: packages/compiler/src/i18n/serializers/xliff2.ts [hunk 4]
```diff
@@ -173,7 +173,12 @@ class _WriteVisitor implements i18n.Visitor {
   }
 
   visitIcuPlaceholder(ph: i18n.IcuPlaceholder, context?: any): xml.Node[] {
-    return [new xml.Tag(_PLACEHOLDER_TAG, {id: (this._nextPlaceholderId++).toString()})];
+    const cases = Object.keys(ph.value.cases).map((value: string) => value + ' {...}').join(' ');
+    return [new xml.Tag(_PLACEHOLDER_TAG, {
+      id: (this._nextPlaceholderId++).toString(),
```
Comment (vicb) [inline packages/compiler/src/i18n/serializers/xliff2.ts]: move the `this._nextPlaceholderId++` outside of the return value.

Comment (vicb) [inline packages/compiler/src/i18n/serializers/xliff2.ts]: _(update the code everywhere this applies. thx)_

File: packages/compiler/test/i18n/serializers/xliff2_spec.ts [hunk 5]
```diff
@@ -23,7 +23,7 @@ const HTML = `
 <p i18n="ph names"><br><img src="1.jpg"><img src="2.jpg"></p>
 <p i18n="empty element">hello <span></span></p>
 <p i18n="@@baz">{ count, plural, =0 { { sex, select, other {<p>deeply nested</p>}} }}</p>
-<p i18n>{ count, plural, =0 { { sex, select, other {<p>deeply nested</p>}} }}</p>
+<p i18n>Test: { count, plural, =0 { { sex, select, other {<p>deeply nested</p>}} } =other {a lot}}</p>
```
Comment (vicb) [inline packages/compiler/test/i18n/serializers/xliff2_spec.ts]: why this change ?

File: packages/compiler/test/i18n/integration_xliff_spec.ts [hunk 6]
```diff
@@ -73,6 +73,14 @@ const XLIFF_TOMERGE = `
         <source><x id="START_ITALIC_TEXT" ctype="x-i"/>with placeholders<x id="CLOSE_ITALIC_TEXT" ctype="x-i"/></source>
         <target><x id="START_ITALIC_TEXT" ctype="x-i"/>avec des espaces réservés<x id="CLOSE_ITALIC_TEXT" ctype="x-i"/></target>
       </trans-unit>
+      <trans-unit id="651d7249d3a225037eb66f3433d98ad4a86f0a22" datatype="html">
```
Comment (vicb) [inline packages/compiler/test/i18n/integration_xliff_spec.ts]: why in this PR ?

File: packages/compiler/test/i18n/serializers/xliff2_spec.ts (hunk 5 above)
Comment (ocombe) [inline packages/compiler/test/i18n/serializers/xliff2_spec.ts]: To have a test with placeholders ICUs as well, to make sure that it works with all functions of the Xliff2 WriteVisitor

File: packages/compiler/src/i18n/serializers/xliff2.ts (hunk 4 above)
Comment (ocombe) [inline packages/compiler/src/i18n/serializers/xliff2.ts]: like I said previously, it doesn't work if I do that because the ids will start at 1 instead of 0, unless I initialize it at -1 if you want (seems weird)

File: packages/compiler/test/i18n/integration_xliff_spec.ts (hunk 6 above)
Comment (ocombe) [inline packages/compiler/test/i18n/integration_xliff_spec.ts]: I needed to check the id generation with nested placeholders that have the same tag to make sure that it generates valid code, and since I changed it in integration_common_spec, it impact all the other integration tests as well
